lib/application_group.ts

```diff
@@ -0,0 +1,224 @@
+import { RawStorageKey } from '@Lib/storage_keys';
+import { removeFromArray, findInArray } from '@Lib/utils';
+import { UuidString, DeinitSource } from './types';
+import { SNApplication } from './application';
+import { PureService } from '@Services/pure_service';
+import { DeviceInterface } from '@Lib/device_interface';
+import { Uuid } from '@Lib/uuid';
+
+export type ApplicationDescriptor = {
+  identifier: string | UuidString,
+  // userUuid?: UuidString,
+  label: string,
+  /** Whether the application is the primary user-facing selected application */
+  primary: boolean,
+}
+
+export type DescriptorRecord = Record<string, ApplicationDescriptor>
+
+type AppGroupCallback = {
+  applicationCreator: (
+    descriptor: ApplicationDescriptor,
+    deviceInterface: DeviceInterface
+  ) => SNApplication
+}
+
+type AppGroupChangeCallback = () => void
+
+export class SNApplicationGroup extends PureService {
+  public primaryApplication?: SNApplication
+  private descriptorRecord!: DescriptorRecord
+  private changeObservers: AppGroupChangeCallback[] = []
+  callback!: AppGroupCallback
+  private applications: SNApplication[] = []
+
+  constructor(public deviceInterface: DeviceInterface) {
+    super();
+  }
+
+  deinit() {
+    super.deinit();
+    this.deviceInterface.deinit();
+    (this.deviceInterface as any) = undefined;
+  }
+
+  public async initialize(callback: AppGroupCallback) {
+    this.callback = callback;
+    this.descriptorRecord = (
+      await this.deviceInterface!.getJsonParsedRawStorageValue(RawStorageKey.DescriptorRecord)
+    ) as DescriptorRecord;
+    if (!this.descriptorRecord) {
+      await this.createDescriptorRecord();
+    }
+    const primaryDescriptor = this.findPrimaryDescriptor();
+    if (!primaryDescriptor) {
+      throw Error('No primary application descriptor found. Ensure migrations have been run.')
+    }
+    const application = this.buildApplication(primaryDescriptor);
+    this.applications.push(application);
+    this.setPrimaryApplication(application, false);
+  }
+
+  private async createDescriptorRecord() {
+    const descriptorRecord = {} as DescriptorRecord;
+    const descriptor = {
+      /** The identifier 'standardnotes' is used because this was the database name of
+       * Standard Notes web/desktop */
+      identifier: "standardnotes",
+      label: 'Main Application',
+      primary: true
+    } as ApplicationDescriptor;
+    descriptorRecord[descriptor.identifier] = descriptor;
```

Comment:
why not initialize descriptorRecord here so you don't have to cast it (destructive operation)? i-e:
```ts
const descriptorRecord: DescriptorRecord = { [descriptor.identifier]: descriptor };
```